serve/mlc_serve/model/model_common.py

```diff
@@ -155,17 +155,22 @@ def _is_safe_to_sample(prob_like):
         torch.cuda.nvtx.range_pop()
         return None
 
-    res_random = torch.multinomial(probs, 1, True).cpu().numpy()[:, 0]
+    res_random = torch.multinomial(probs, 1, True)[:, 0].cpu().numpy()
 
     if logits_random.shape[0] == num_seq:
         torch.cuda.nvtx.range_pop()
         return res_random
 
+    mask_random_cpu = torch.tensor(
+        [p.sampling_type == SamplingType.RANDOM for p in sampling_params],
+        dtype=torch.bool
+    )
+    mask_greedy_cpu = torch.logical_not(mask_random_cpu)
```

Comment:
When you create a GPU mask, PT will first create a CPU mask and do `cudaMemcpy` under the hood. So you can create a CPU mask once, and create a GPU mask by copying the CPU mask explicitly. So you can avoid creating a CPU mask twice.

Reply:
If you use greedy case you do not need CPU mask at all

Reply:
Please, recheck